### PR TITLE
[iOS] Fix NOTIFICATION permission prior SDK32

### LINKED ISSUE #3458: await Permissions.askAsync(Permissions.NOTIFICATIONS) does not return on first call for all supported SDKs (with Snack)
Snack: https://snack.expo.io/ByRNCO5EE
This happens for all SDK versions supported by Snack except 32.

### Steps to Reproduce

1. Ask user for permissions to send them notifications with `const result = await Permissions.askAsync(Permissions.NOTIFICATIONS)`
2. User is prompted to allow/not allow notifications
3. User taps "Allow" or "Don't Allow"
4. Nothing happens. No error, and JS execution ends.

### Expected Behavior

User taps an option and the async function is resolved, returning the user's response and continuing execution.

### Actual Behavior

User taps a response and it fails quietly.
If the call is made a second time it works successfully! This is not ideal for obvious reasons.

#### My current diagnostics (not sure they're relevant):

  Expo CLI 2.10.0 environment info:
    System:
      OS: macOS 10.14
      Shell: 3.2.57 - /bin/bash
    Binaries:
      Node: 10.15.1 - /usr/local/bin/node
      Yarn: 1.13.0 - /usr/local/bin/yarn
      npm: 6.4.1 - /usr/local/bin/npm
    IDEs:
      Xcode: /undefined - /usr/bin/xcodebuild
    npmPackages:
      expo: ^28.0.0 => 28.0.1 
      react: 16.3.1 => 16.3.1 
      react-native: https://github.com/expo/react-native/archive/sdk-28.0.0.tar.gz => 0.55.4 
      react-navigation: ~2.3.1 => 2.3.1 
    npmGlobalPackages:
      expo-cli: 2.10.0

Targeting iOS

### PR DESCRIPTION
**Pull request by bbarthec**
# Why

Resolve #3458

# How

Restore removed deprecated `AppDelegate` callbacks responsible for triggering `Permission.NOTIFICATION` resolving logic prior SDK32.

# Test Plan

`ncl` PermissionsScreen with `NOTIFICATION` permission using SDK lower than 32.

### CODE REVIEW

File: ios/Client/EXAppDelegate.m [hunk 1]
```diff
@@ -157,6 +157,12 @@ - (void)application:(UIApplication *)application didReceiveRemoteNotification:(N
   [(EXTaskService *)[EXModuleRegistryProvider getSingletonModuleForClass:EXTaskService.class] runTasksWithReason:EXTaskLaunchReasonRemoteNotification userInfo:userInfo completionHandler:completionHandler];
 }
 
+// do not remove until SDK31 is dropped
```
Comment (sjchmiela) [inline ios/Client/EXAppDelegate.m]:
```suggestion
// TODO: Remove once SDK31 is phased out
```

File: ios/Exponent/ExpoKit/ExpoKit.h [hunk 2]
```diff
@@ -52,6 +52,8 @@ FOUNDATION_EXPORT NSString * const EXAppDidRegisterUserNotificationSettingsNotif
 - (void)application:(UIApplication *)application didFinishLaunchingWithOptions:(nullable NSDictionary *)launchOptions;
 
 #pragma mark - APNS hooks
+// do not remove until SDK31 is dropped
```
Comment (sjchmiela) [inline ios/Exponent/ExpoKit/ExpoKit.h]:
```suggestion
// TODO: Remove once SDK31 is phased out
```

File: ios/Exponent/ExpoKit/ExpoKit.m [hunk 3]
```diff
@@ -149,6 +149,12 @@ - (void)application:(UIApplication *)application didReceiveRemoteNotification:(N
   completionHandler(UIBackgroundFetchResultNoData);
 }
 
+// do not remove until SDK31 is dropped
```
Comment (sjchmiela) [inline ios/Exponent/ExpoKit/ExpoKit.m]:
```suggestion
// TODO: Remove once SDK31 is phased out
```

**Review by sjchmiela — APPROVED** [review]
Thanks for taking care of this! 👍

**Review by Szymon20000 — APPROVED** [review]
Looks good!

**Comment by jSkrills** [comment]
@bbarthec is this in a certain version of expo? Where can I see that from here?

**Comment by bbarthec** [comment]
@jSkrills upcoming SDK33 release will contain this fix 🙂

**Comment by jSkrills** [comment]
@bbarthec I am still seeing the unwanted behavior of #3458 and #1354 after upgrading to sdk 33. Other permissions seem to be resolving fine but notifications still never returns. I do have a detached app but I don't think that should make a difference. I followed all the sdk upgrade instructions to 33 as well as updating ExpoKit to sdk 33 as well.

**Comment by fabiosarcansky** [comment]
@jSkrills were you able to resolve it? 
I'm working with SDK 32 and I have to deploy a new release soon. I don't want to spend a lot of time upgrading to SDK 33 for nothing.
Thanks!

**Comment by jSkrills** [comment]
@fabiosarcansky No I haven't resolved it yet. I was hoping for a response from @bbarthec. I thought maybe I have things configured wrong and that's why I'm still seeing the behavior.

**Comment by jSkrills** [comment]
@bbarthec Any thoughts on why I might be still seeing this behavior after upgrading?

**Comment by bbarthec** [comment]
@jSkrills, I'm going to check what might be wrong regarding these two issues you've mentioned (#3458 and #1354).

**Comment by jSkrills** [comment]
@bbarthec Ok great, thank you so much. Let me know if I can provide you any helpful info in your troubleshooting.

**Comment by bbarthec** [comment]
@jSkrills, I've just confirmed that this [snack](snack.expo.io/ByRNCO5EE) (taken from #3458) is working as expected with SDK33:
- [x] in Expo Client.
- [x] as ejected Expo project launched via XCode.
- [x] as standalone iOS application built by `turtle` and published on TestFlight.

**Comment by jSkrills** [comment]
Hey @bbarthec I am seeing this behavior on the second configuration with the ejected app launched via XCode. I had my import setup differently than your snack bc I was following the sdk 33 docs. I had it setup like this: import * as Permissions from 'expo-permissions'. When I try it how you have it in the snack I DO get a response back after I decide to allow notifications but it is this response: 

The following APIs have moved to separate packages and importing them from the "expo" package is deprecated: Permissions.

1. Add correct versions of these packages to your project using:

   expo install expo-permissions

   If "install" is not recognized as an expo command, update your expo-cli installation.

2. Change your imports so they use specific packages instead of the "expo" package:

 - import { Permissions } from 'expo' -> import * as Permissions from 'expo-permissions'

Basically telling me I shouldn't import the permissions the way you are in the snack. I don't know why you don't receive that same message with sdk 33 though.

**Comment by bbarthec** [comment]
@jSkrills, we're moving from `import { Package } from 'expo'` towards `import * as Package from 'expo-package'`, but in SDK 33 we're supporting both (I suggest you use the latter).
Just to clarify, are you receiving expected results calling `Permissions.askAsync`?

**Comment by jSkrills** [comment]
@bbarthec OK, sounds good. No, I am not receiving the expected results that call, specifically just with notifications though like this: 
`response = await Permissions.askAsync(Permissions.NOTIFICATIONS);`

Basically once I hit this line the modal pops up asking the user but then when I answer allow or don't allow I can never see the response.

Also to note:
`response = await Permissions.askAsync(Permissions.CAMERA);` 
and...
`response = await Permissions.askAsync(Permissions.CAMERA_ROLL);`

work perfectly fine in the same component as well as getAsync() for all those permissions.

**Comment by bbarthec** [comment]
@jSkrills, I'm having difficulties reaching scenario that does not work 😢 I'd like to ask you whether you could provide some mvp repository covering the issue? That would be a huge help for me as I'm willing to solve this issue as fast as possible 🙏

**Comment by jSkrills** [comment]
@bbarthec, Yeah sure I should have some time to try and knock that out tonight. I'll let you know when it's done. Thank you a ton for your help and perseverance so far btw, it really means a lot.

**Comment by jSkrills** [comment]
@bbarthec  I added you as a collaborator on a private repo and documented the steps I took and the output I am seeing in the readme. 

The only difference between what I saw on the mvp demo app and what I see on my for real app is that my app will never return anything after the console outputs "asking" So it will ask the user to allow or don't allow permissions then after I answer the console never logs anything again. The mvp demo app will always return me a status of undetermined even when I answer allow.

However I think if we are able to get to the bottom of one we'll get to the bottom of both.

**Comment by jSkrills** [comment]
@bbarthec Have you had a chance to check out that repo yet?

**Comment by bbarthec** [comment]
@jSkrills, unfortunately I'm busy right now by SDK 34 releasing process. I will jump back to this issue right after new SDK is out. I'm sorry for this inconvenience 😔

**Comment by jSkrills** [comment]
@bbarthec ok, I get that, please just keep me posted.

**Comment by jSkrills** [comment]
@bbarthec Have you had a minute to checkout that mvp repo I made yet?

**Comment by bbarthec** [comment]
@jSkrills, I've just checked your repository and obtained logs (very same as you described in your MVP repo). Let's take a look at them:
```
hello from addPattern
HomeScreen asked before:  true
existing status:  Object { <-- these are results from native method
  "expires": "never",
  "permissions": Object {
    "notifications": Object {
      "allowsAlert": true,
      "allowsBadge": true,
      "allowsSound": true,
      "expires": "never",
      "status": "undetermined",
    },
  },
  "status": "undetermined",
}
asking
finalStatus:  undetermined
HomeScreen finalStatus:  undetermined
returning false
checkPermission notification response is not truthy :/
```

Native method actually returned (almost) correct results. Let's take a look at [docs](https://docs.expo.io/versions/v33.0.0/sdk/permissions/#permissionsnotifications) and especially this small subsection:
> **Note: On iOS, asking for this permission asks the user not only for permission to register for push/remote notifications, but also for showing notifications as such. At the moment remote notifications will only be received when notifications are permitted to play a sound, change the app badge or be displayed as an alert. As iOS is more detailed when it comes to notifications permissions, this permission status will contain not only status and expires, but also Boolean values for allowsSound, allowsAlert and allowsBadge.

As we can see, results returned from native side contain `allows***`, each with `true` value. Therefore we do have access to `NOTIFICATIONS` 😃.

The problem is why combined `status` is not saying `granted` 🤔 I've created separate issue (https://github.com/expo/expo/issues/5093) for this, but for now I recommend using some js code to ensure the permission is granted, for example like this:
```js
if (permissionType = Permission.NOTIFICATIONS) {
  const { [Permission.NOTIFICATIONS]: { allowsAlert, allowsBadge, allowsSound } } = await Permissions.askAsync(Permission.NOTIFICATIONS);
  return allowsAlert && allowsBadge && allowsSound;
}
```

**Comment by jSkrills** [comment]
@bbarthec It's good to know I am not going crazy lol. Ok I will keep an eye on the other issue and move forward with the workaround on iOS, thank you so much for your help so far and for not forgetting about me.

**Comment by jSkrills** [comment]
@bbarthec While your workaround would work if I was ever getting a response in my actual app this call: `response = await Permissions.askAsync(Permissions.NOTIFICATIONS);` still will never return me anything so it doesn't do me much good. 

In diff'ing the two projects I am actually setting up the PushNotificationIOS stuff in my app documented [here](https://facebook.github.io/react-native/docs/0.59/pushnotificationios#docsNav). Out of curiosity I removed all the suggested setup things from my project in xCode and I was able to then actually get a response back from that expo permissions call.

Do you know why setting up the react native PushNotificationIOS stuff would prevent me from getting a permissions response from expo?

**Comment by jSkrills** [comment]
Hey @bbarthec have you had a chance to check this out at all?

**Comment by bbarthec** [comment]
@jSkrills, right now `expo-permissions` is strongly coupled with notifications module from Expo. Therefore there's no way to use `expo-permissions` with external notifications libraries/implementations 😢 

Do not be discouraged though! Right now there two major refactors going on:
- one aiming to refactor `expo-permissions` #5061
- one aiming to refactor `expo-notifications` #4480

These two would tremendously simplify using these two modules and achieving working scenarios, even in bare flow. 🎉

**Comment by jSkrills** [comment]
@bbarthec Well that would explain my struggles, I am glad to at least have some clarity. Is that clearly documented anywhere and I just overlooked it? I thought that I read the docs of things over several times and don't remember that standing out. Because if it's not I feel it should be, to save someone else the hassle and troubleshooting heartache. Also, is there any estimate / wild ass guess on when those refactors would be available?

**Comment by jSkrills** [comment]
@bbarthec any thoughts ^ amigo?

**Comment by bbarthec** [comment]
@jSkrills, following docs: [`This API is pre-installed in managed apps. It is not available for bare React Native apps`](https://docs.expo.io/versions/latest/sdk/notifications/#installation) - the only way to use `notifications` (at the time of writing this comment - SDK35) is to use it in `managed flow`.
`bare` flow would start supporting it when it's successfully extracted from monolithic `ExpoKit` into `unimodule`.

Mentioned refactors are constantly being improved and are expected to be merged in one/two release cycles and that would mean in a month or two. I'm sorry that it's taking so long .. 🙇
